### ISSUE TITLE
Enable crash logging in `Networking` layer

### ISSUE THREAD

**Issue by jaclync** [woocommerce/woocommerce-ios#2372]
Now that we are enabling crash logging for a top crash in `Storage` layer https://github.com/woocommerce/woocommerce-ios/issues/2364, let's also add it to the `Networking` layer. Maybe we can start logging where we have `DDLogError` in `Networking` (like when it cannot map the response in `Remote`) @woocommerce/ios-developers ?

Also, since `Networking` and `Storage` are independent frameworks, I was wondering if it's worth adding a base framework where some shared protocols/helpers can be in without any dependencies?

**Comment by shiki** [comment]
While reviewing #2366, I thought about using `DDLogError` instead of `SentryCrashLogger` directly. Are you proposing we **only use** `DDLogError` for both `Storage` and `Networking`? I feel like we should merge these logging APIs into one. I'm just not sure which one is better. 🤔 

What do you think? 

> Maybe we can start logging where we have DDLogError in Networking (like when it cannot map the response in Remote)

That makes sense to me. 

> I was wondering if it's worth adding a base framework where some shared protocols/helpers can be in without any dependencies?

Well, this has always been the plan. I just haven't gotten to it yet. 😄 Feel free to take it though!

**Comment by jaclync** [comment]
> While reviewing #2366, I thought about using `DDLogError` instead of `SentryCrashLogger` directly. Are you proposing we **only use** `DDLogError` for both `Storage` and `Networking`? I feel like we should merge these logging APIs into one. I'm just not sure which one is better. 🤔

From my understanding, `DDLogError` is only logged to the application log that is included in a support ticket, is it correct? For logging in https://github.com/woocommerce/woocommerce-ios/pull/2366, I think logging to Sentry would be more helpful for getting error information from all the users (if they opted in) than `DDLogError`.

Maybe we could define a few levels of logging like below, feel free to edit directly or suggest edits @woocommerce/ios-developers !

severity level | where events are sent | details
-- | -- | --
fatal (since the debug message in `fatalError` isn't recorded) | Sentry | when the app cannot recover from the current state and should terminate
non-fatal error | Sentry | when the app encounters an error that deserves our attention
error | `DDLogError` | when the app encounters an error that could be helpful when they submit a support ticket
warning | `DDLogWarn` | when the app receives a warning that could be helpful when they submit a support ticket
info (higher priority) | Sentry | when the app is in a scenario that isn't a warning nor an error but deserves our attention
info | `DDLogInfo` | when the app is in a scenario that isn't a warning nor an error but could be helpful when they submit a support ticket
debug | Sentry | I haven't thought of a good use case for this now 🤔

**Comment by shiki** [comment]
> From my understanding, DDLogError is only logged to the application log that is included in a support ticket, is it correct?

Yes, currently. IIRC, we can configure CocoaLumberjack to capture `DDLogError` and pass them through Sentry. With this, `DDLogError` will be similar to `SentryCrashLogger.logMessage()`. That's when it can be confusing because we'll have to decide which one to use while writing code. 😬 

That's why I was thinking we could consolidate all of these so that we only have one “logging API”. Maybe something like:

```swift
enum Destination {
    // Just log file
    case log
    // Both log file and Sentry
    case sentry
}

func logInfo(message: String, destination: Destination = .log) {

}

func logError(message: String, destination: Destination = .sentry) {

}
```

I'm sure there's a much cleaner way to do that. ☝️ 😄 

---

Re: Log Level Tables you shared:

It can be tricky to decide whether something is a "non-fatal error" and "error". I guess that has to be decided by the developer, right? So, using the sample code above, one could state the `destination` to where the log should go.

**Comment by shiki** [comment]
We should also make sure that we will be able to **act** on the event. For example, if we are submitting a parse error event to Sentry, we should make sure that the message includes why the parse failed. 

The Platform team also suggested that we should set up a Sentry alert for these events so that they don't get lost in the noise. 

Ref: pbMoDN-y8-p2#comment-939

**Comment by jaclync** [comment]
Moved to Linear for follow-up: https://linear.app/a8c/issue/WOOMOB-3466/route-yosemite-and-networking-errors-to-crash-loggingsentry

Closing this GitHub issue as superseded by the consolidated Linear ticket.
